ibtk/src/utilities/IBTKInit.cpp

```diff
@@ -0,0 +1,79 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2014 - 2019 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "ibtk/IBTKInit.h"
+#include "ibtk/IBTK_MPI.h"
+#include "ibtk/app_namespaces.h"
+
+namespace IBTK
+{
+/////////////////////////////// STATIC //////////////////////////////////////
+#ifdef IBTK_HAVE_LIBMESH
+LibMeshInit* IBTKInit::s_libmesh_init = nullptr;
+#endif
+bool IBTKInit::s_initialized = false;
+
+void
+IBTKInit::initialize(int argc, char** argv, IBTK_MPI::comm communicator, char* petsc_file, char* petsc_help)
+{
+    if (s_initialized)
+    {
+        TBOX_WARNING("IBAMR already initialized. Multiple initialize() calls will be ignored.");
+    }
+    static IBTKInit instance(argc, argv, communicator, petsc_file, petsc_help);
+}
+
+/////////////////////////////// PRIVATE //////////////////////////////////////
+
+IBTKInit::IBTKInit(int argc, char** argv, IBTK_MPI::comm communicator, char* petsc_file, char* petsc_help)
+{
+#ifdef IBTK_HAVE_LIBMESH
+    s_libmesh_init = new LibMeshInit(argc, argv, communicator);
```

Comment:
We (perhaps only me) sometimes use this in programs - we should make it returnable by reference.

Reply:
It's returnable by pointer. Would reference be better?

Reply:
I missed that function in the header - oops.

I would prefer to have it returned by reference.

Reply:
Would it be possible to use pointers / smart pointers in a way that makes access to `LibMeshInit` safer after shutting things down?

Reply:
Yes  - we did this a while back in deal.II with the `SmartPointer` class. essentially, one must also store a boolean indicating whether or not the pointed-to object is alive.

Reply:
Alternatively, we could make `LibMeshInit` a member of this class instead of statically allocating it - that is probably the better solution.

Reply:
That seems best to me -- seems like it avoids most of the de-initialization issues that I am imagining.

Reply:
Yes. In particular, the only static member of this class should be a boolean storing whether or not we have already created it.